#164139695 Add search and filter functionality

codinger41 opened the pull request:
#### What does this PR do?
Adds the feature for users to be able to search and filter articles.
#### Description of Task to be completed?
Users should be able to search for articles. While searching, users should be able to filter their search results. 
This new endpoint takes these queries:
- `term` the search keyword
- `categoryId` the category the user wants to search from. We'll provide this based on what the user selects on the front end.
- `author` the username or name of the author they want to search from.
- `startDate` and `endDate` the start and end dates for a date range filter
- `tags` a string of tags they want to search from.
#### How should this be manually tested?
- run  `npm run db-migrate` to get all database migration updates.
- make a GET request to the `/search` route with the above query parameters.
#### Any background context you want to provide?
- N/A
#### What are the relevant pivotal tracker stories?
https://www.pivotaltracker.com/story/show/164139695

#### Screenshots (if appropriate)
#### Questions:
N/A

kleva-j committed:
feat(create_article): add functionality to create articles
- setup middlewares to validate article details and to authenticate user token
- create database models for articles
- setup routes for the endpoint to post an article
- write test for testing the functionality to create an article
- add documentation for this endpoint
[Delievers #164139686]

kleva-j committed:
ft(add_image): enable image upload for article
- add middlewares to generate slug for articles
- create functions to validate article details
- update migrations file for articles
- create middleware to add and upload images
[Delivers #164139686]

kleva-j committed:
ft(add_image): enable image upload for article
- add middlewares to generate slug for articles
- create functions to validate article details
- update migrations file for articles
- create middleware to add and upload images
[Delivers #164139686]

codinger41 committed:
merge article-creation

codinger41 committed:
resolve merge conflicts

codinger41 committed:
resolve merge conflicts

codinger41 committed:
feat(search): implement search and filter functionality
- adds a route for searching for articles
- cater for search filtering with tags, author name, and date range

[Finishes #164139695]

chingsley committed:
feat(create-category):  creating categories

- develop algorithm for creating categories
- write tests for the endpoint
- create the validation function to validate inputs
- create the controller function to perform the action
- ensure all tests are passing

[Finishes #164603412]

codinger41 committed:
feat(search): implement search and filter functionality
- adds a route for searching for articles
- cater for search filtering with tags, author name, and date range

[Finishes #164139695]

codinger41 committed:
feat(search): implement search and filter functionality
- adds a route for searching for articles
- cater for search filtering with tags, author name, and date range

[Finishes #164139695]

olajide1234 commented on controllers/articles.js:
```diff
@@ -34,4 +40,119 @@ export default class ArticleController {
+        success: true
+      });
+    } catch (error) {
+      return res.status(200).json({
```
Please consider returning a 404 here.

Dubby20 commented on controllers/articles.js:
```diff
@@ -34,4 +40,118 @@ export default class ArticleController {
+    } catch (error) {
+      return res.status(404).json({
+        success: false,
+        errors: ['Article not found.'],
```
This should be an error, not errors

leksyib replied:
Our convention uses `errors` and it's an array. If I change it to `error`, we won't be consistent and when we're working on the front end, we won't know the exact one to use. Let's keep being consistent so it'll be easy for us. Thanks.

Dubby20 reviewed (approved):
Looks good to merge

codinger41 committed:
feat(search): implement search and filter functionality
- adds a route for searching for articles
- cater for search filtering with tags, author name, and date range

[Finishes #164139695]

codinger41 committed:
feat(search): implement search and filter functionality
- adds a route for searching for articles
- cater for search filtering with tags, author name, and date range

[Finishes #164139695]

codinger41 committed:
feat(search): implement search and filter functionality
- adds a route for searching for articles
- cater for search filtering with tags, author name, and date range

[Finishes #164139695]
